src/borg/testsuite/repo_new.py

```diff
@@ -0,0 +1,257 @@
+import os
+
+import pytest
+
+from ..helpers import IntegrityError
+from ..repository import Repository, MAX_DATA_SIZE, TAG_DELETE, TAG_PUT2, TAG_PUT, TAG_COMMIT
+from ..repoobj import RepoObj
+from .hashindex import H
+
+
+UNSPECIFIED = object()  # for default values where we can't use None
+
+
+def fchunk(data, meta=b""):
+    # create a raw chunk that has valid RepoObj layout, but does not use encryption or compression.
+    meta_len = RepoObj.meta_len_hdr.pack(len(meta))
+    assert isinstance(data, bytes)
+    chunk = meta_len + meta + data
+    return chunk
+
+
+def pchunk(chunk):
+    # parse data and meta from a raw chunk made by fchunk
+    meta_len_size = RepoObj.meta_len_hdr.size
+    meta_len = chunk[:meta_len_size]
+    meta_len = RepoObj.meta_len_hdr.unpack(meta_len)[0]
+    meta = chunk[meta_len_size : meta_len_size + meta_len]
+    data = chunk[meta_len_size + meta_len :]
+    return data, meta
+
+
+def pdchunk(chunk):
+    # parse only data from a raw chunk made by fchunk
+    return pchunk(chunk)[0]
+
+
+class RepositoryTestCaseBase:
+    def reopen(self, repository, exclusive=UNSPECIFIED):
+        repository.__exit__(None, None, None)
+        repository.__enter__()
+        repository.exclusive = exclusive
+
+    def add_keys(self, repository):
+        repository.put(H(0), fchunk(b"foo"))
+        repository.put(H(1), fchunk(b"bar"))
+        repository.put(H(3), fchunk(b"bar"))
+        repository.commit(compact=False)
+        repository.put(H(1), fchunk(b"bar2"))
+        repository.put(H(2), fchunk(b"boo"))
+        repository.delete(H(3))
+
+    def repo_dump(self, repository, label=None):
+        label = label + ": " if label is not None else ""
+        H_trans = {H(i): i for i in range(10)}
+        H_trans[None] = -1  # key == None appears in commits
+        tag_trans = {TAG_PUT2: "put2", TAG_PUT: "put", TAG_DELETE: "del", TAG_COMMIT: "comm"}
+        for segment, fn in repository.io.segment_iterator():
+            for tag, key, offset, size, _ in repository.io.iter_objects(segment):
+                print("%s%s H(%d) -> %s[%d..+%d]" % (label, tag_trans[tag], H_trans[key], fn, offset, size))
+        print()
+
+
+class TestRepository(RepositoryTestCaseBase):
+    @pytest.fixture()
+    def repository(self, tmpdir):
+        self.repository_location = os.path.join(str(tmpdir), "repository")
+        with Repository(self.repository_location, exclusive=True, create=True) as repository:
+            yield repository
+
+    def test1(self, repository):
+        for x in range(100):
```

Comment:
Guess we should rename these tests to something more descriptive?

Reply:
sure.